### DISCUSSION TITLE
Compact data type system

### DISCUSSION THREAD

**Discussion by mbasaglia** [lottie/lottie-spec#22] [Ideas]
Just an idea to represent value types as a compact string. I think this is useful for declaring slots, could be useful for future extensions too.

Basically representing types property types with a short string that helps parsers know what types are being handled. My idea is to have the following fields encoded in the string:

* Base type
* Length (optional)
* Modifiers (optional)

_Base type_ could have the following values:

* `v` for vector
*  `s` for scalar
*  `c` for color
*  `p` for bezier path
* `g` for gradient

_Length_ is the number of components for vectors and the like

_Modifiers_ give more context, examples:

* `p` percentage, like scale, opacity where a value of 100 means "100%". This is useful as implementation might want to convert it to [0-1] values internally
* `a` angle, for rotations and the like
* `t` transparency, for RGB + A gradients

Examples:

* `v2` 2D vector data, eg: position
* `v2p` 2D "percentage" vector for scale
* `sa` Scalar angle (rotation)
* `c3` RGB color, (the 3 could be optional)
* `c4` RGBA color
* `g` RGB Gradient (I don't think we need to specify the size here)
* `gt` RGB + Transparency Gradient

Mainly this would be used to specify slot types to make them easier to parse but could also be used in expressions and effects if we want to rework those systems.

**Comment by fmalita** [comment]
Currently, slots use an [integral type enum](https://github.com/bodymovin/bodymovin-extension/blob/4f1de8ebfe5beaebc4c7d54fbbdf363ca685478c/bundle/jsx/utils/essentialPropertiesHelper.jsx#L15).

The advantage of this proposal is composability and better scalability.  OTOH the integral enum approach is slightly easier to parse and more in line with other existing Lottie type specifiers.

Do you expect a lot of type variations, or a fairly stable/small set?  If the latter, expanding the existing enum might be more straightforward.
